envoy: Pass nil completion if acks are not expected.

jrajahalme opened the pull request:
Completions are cleaned up when an ack or nack for the resource type
is received. This never happens if there are no configured L7 proxies.

Prevent indefinite collection of completions in this case by passing
nil completion to Upsert().

Fixes: 99a73a2fbd ("envoy: Update revision upon acks from NPDS")
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5911)
<!-- Reviewable:end -->

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -194,10 +196,12 @@ func (m *AckingResourceMutatorWrapper) Delete(typeURL string, resourceName strin
-			logfields.XDSResourceName: resourceName,
-		}).Fatal("no completion given to delete xDS resource")
+		// No completion in upsert, revert also without completion
+		return func(_ *completion.Completion) {
```
I think it would be simpler and more consistent to just use with the revert func below. Even if this mutation didn't have a completion, it doesn't mean that you should ignore any completion for reverting it.
Just remove this whole `if` block here, and enclose the stuff below into an `if c != nil`:
```
if c != nil {
    if _, found := m.pendingCompletions[c]; found {
        ...
    }
    m.addDeleteCompletion(typeURL, version, nodeIDs, c)
}
```

rlenglet commented on pkg/envoy/xds/ack.go:
```diff
@@ -139,15 +139,17 @@ func (m *AckingResourceMutatorWrapper) Upsert(typeURL string, resourceName strin
-			logfields.XDSResourceName: resourceName,
-		}).Fatal("no completion given to upsert xDS resource")
+		// No completion in upsert, revert also without completion
+		return func(_ *completion.Completion) {
```
I think it would be simpler and more consistent to just use with the revert func below. Even if this mutation didn't have a completion, it doesn't mean that you should ignore any completion for reverting it.
Just remove this whole `if` block here, and enclose the stuff below into an `if c != nil`:
```
if c != nil {
    if _, found := m.pendingCompletions[c]; found {
        ...
    }
    ...
    m.pendingCompletions[c] = comp
}
```

jrajahalme replied:
Right, will make the original revert funcs deal with nil completions as well.

rlenglet replied:
Cool, I didn't catch that either.

jrajahalme committed:
envoy: Pass nil completion if Acks are not expected.

Completions are cleaned up when an ack or nack for the resource type
is received. This never happens if there are no configured L7 proxies.

Prevent indefinite collection of completions in this case by passing
nil completion to Upsert().

Fixes: 99a73a2fbd ("envoy: Update revision upon acks from NPDS")
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

rlenglet reviewed (approved):
Great, thanks!
